src/server/middleware/csrf-token.js

```diff
@@ -0,0 +1,67 @@
+import uuid from 'node-uuid';
+
+export const generateToken = (req, res, next) => {
+  // Don't generate new tokens for API calls
+  if (req.path.match(/^\/api/)) {
+    return next();
+  }
+
+  // TODO: Remove debug
+  const csrfToken = uuid.v4();
+  if (typeof req.session.csrfTokens === 'undefined') {
+    req.session.csrfTokens = [];
+  }
+
+  req.session.csrfTokens.push(csrfToken);
+
+  // Only 5 valid CSRF tokens at a time. Safe to increase number.
+  while (req.session.csrfTokens.length > 5) {
+    req.session.csrfTokens.shift();
+  }
```

Comment:
The other thing you could consider is generating CSRF tokens using HMAC based on a server-side secret and the session ID.  That way the server doesn't need to store the tokens, and they're still unforgeable as long as the secret stays secret and the underlying hash algorithm is sound.  You can include additional criteria like an expiration timestamp if you like.  It's the same kind of idea as macaroons.

Web-searching for "CSRF HMAC" finds a few things about this, e.g. [hmac-csrf](https://www.npmjs.com/package/hmac-csrf), though I haven't looked them over in any detail.